velox/functions/prestosql/tests/ArraySortTest.cpp

```diff
@@ -405,6 +405,54 @@ TEST_P(ArraySortTest, constant) {
   assertEqualVectors(expected, result);
 }
 
+TEST_P(ArraySortTest, multilevelDictionaries) {
+  if (GetParam() != TypeKind::BIGINT) {
+    GTEST_SKIP() << "Skipping constant test for non-bigint type";
+  }
+
+  // Create dictionaries that are multiple levels deep. Odd numLevels will
+  // return a reverse element and even will return the original elements in
+  // order when decoded.
+  auto multiLevelDictionary = [&](uint32_t numLevels, VectorPtr rootElements) {
+    VectorPtr dictionaryVector = rootElements;
+    for (auto i = 0; i < numLevels; i++) {
+      dictionaryVector = wrapInDictionary(
+          makeIndicesInReverse(rootElements->size()), dictionaryVector);
+    }
+    return dictionaryVector;
+  };
+
+  auto evaluateSort = [&](const VectorPtr& arrayVector,
+                          const VectorPtr& expected) {
+    auto result = evaluate("array_sort(c0)", makeRowVector({arrayVector}));
+    assertEqualVectors(expected, result);
+  };
+
+  auto rootElementVector = makeNullableFlatVector<int64_t>({3, 1, 2});
```

Comment:
Interesting bug, looks like the original code assumes that the elements vector will always be flat, so any kind of dict layer manipulation would not work. for eg, the following test case would throw as it would attempt to copy 5 values whereas the base of the decodedElements only has two.

```
TEST_P(ArraySortTest, multilevelDictionaries) {
  if (GetParam() != TypeKind::BIGINT) {
    GTEST_SKIP() << "Skipping constant test for non-bigint type";
  }
  auto elements = BaseVector::wrapInDictionary(
      nullptr,
      makeIndices(5, [](auto row) { return 0; }),
      5,
      makeFlatVector<int64_t>({1, 2}));

  // Make a single-row array vector: [1, 1, 1, 1, 1].
  auto arrayVector = makeArrayVector({0}, elements);

  ASSERT_NO_THROW(evaluate("array_sort(c0)", makeRowVector({arrayVector})));
}
```

Reply:
No, I think in the case you outlined it will work, since it will call decoded vector on elements and use its base .